### DISCUSSION TITLE
Agenda for 2022-10-27 BMI Council meeting

### DISCUSSION THREAD

**Discussion by mdpiper** [csdms/bmi#116] [Announcements]
Here's the agenda for this Thursday's Council meeting:

* Extensions to core BMI
* New language bindings
* #99: start RFC process

See you at 1 pm MT.

**Comment by mdpiper** [comment]
## Meeting notes

Log of meeting discussion in the comments below. May be messy, but I'll come back and edit later.

**Reply by mdpiper** [reply]
We should make the Zoom link public & post it here. Meetings should be open to all in community.

**Comment by mdpiper** [comment]
### Extensions to core BMI

Idealized use cases from @gregtucker:
1. BMI will grow. Add to BMI to support new functionality. Broad applicability. But loses "B" in Basic
2. B is Basic. Makes it easier for people to implement. Extensions for HPC, parallel.

**Reply by mdpiper** [reply]
Fred asks if there were initial design principles to guide this?

**Reply by mdpiper** [reply]
Scott and @mcflugen gave some suggestions, including design prinicilple in #105.

**Reply by mdpiper** [reply]
@RolfHut, Scott, and @mcflugen agree.

**Reply by mdpiper** [reply]
"Hollywood Principle"

**Reply by mdpiper** [reply]
Fred and @peckhams suggest variable roles could be added to core. Using *get_var_names* would reduce the number of BMI functions!

**Reply by mdpiper** [reply]
@mcflugen Start this as an extension, then migrate to core BMI.

**Reply by mdpiper** [reply]
@RolfHut Add time as a role, a settable variable, as well.

**Reply by mdpiper** [reply]
@RolfHut Core of BMI should become smaller, not bigger, as we move to BMI 3.0.

**Reply by mdpiper** [reply]
@gregtucker Plugins and conflicts.

**Reply by mdpiper** [reply]
@jdhughes-usgs Has also been thinking about plugins wrt XMI. Everybody can develop plugins. Curate and have a way of making a more official ecosystem.

**Reply by mdpiper** [reply]
@peckhams: BMI is not exclusive: can always add additional private convenience functions.

**Reply by mdpiper** [reply]
@RolfHut We should agree on best practices.

**Reply by mdpiper** [reply]
@gregtucker Summarizing discussion. Extensions as a testing ground, migrating to core. Plus Joe's idea of registered extension.

**Reply by mdpiper** [reply]
@peckhams Frozen code problem.

**Reply by mdpiper** [reply]
@mcflugen Lack of testing BMI along with model

**Reply by mdpiper** [reply]
@hrajagers How to distinguish between registered/non-registered extensions?

**Reply by mdpiper** [reply]
@mcflugen We can implement a `get_bmi_extensions` function.

**Reply by mdpiper** [reply]
@RolfHut We should be careful because it could be messy tryign to figure out what extensions a BMI uses.

**Reply by mdpiper** [reply]
@mdpiper We'll make an issue first with a plan. If people are happy with it, we can implement plan in a PR.

**Reply by mdpiper** [reply]
@peckhams We typically think of explicit time stepping. Any issues with implicit time step?

**Reply by mdpiper** [reply]
@hrajagers This should be OK, they've done this at Deltares.

**Reply by mdpiper** [reply]
@RolfHut Has a PR in with `get_state` and `set_state` functions. Should take care of implicit time step.

**Reply by mdpiper** [reply]
@peckhams Has also implemented get/set state in his NOAA work with Fred.

**Comment by mdpiper** [comment]
### Language extensions

How can we handle new language specifications that we don't know well, like Julia and R?

**Reply by mdpiper** [reply]
@hrajagers It's OK to make a value judgement--fully supported by CSDMS vs community submission. Some marker. It would be a resource issue if we supported all languages, even the ones we don't know. Very active group of community submission wouldf get increased recognition.

**Reply by mdpiper** [reply]
@gregtucker Are language specs part of the standard?

**Reply by mdpiper** [reply]
@hrajagers It would be good to ameliorate differences between competing language specs. Work it out. E.g., with Julia.

**Reply by mdpiper** [reply]
@hrajagers Think of this in terms of coupling. If JS talks only to JS, in a JS framework, does it add anything?

**Reply by mdpiper** [reply]
@peckhams There shouldn't be much wiggle room when translating SIDL to a language.

**Reply by mdpiper** [reply]
@mcflugen We won't be writing tools to support JS in pymt, for example.

**Reply by mdpiper** [reply]
The SIDL is standard; the language specs are interpreted from the SIDL. I used my judgement to make specifications that were close to being natural for a developer as  possible. I was favoring the developer when writing the specs.

**Reply by mdpiper** [reply]
@peckhams We should support as many languages as possible. More languages increase strength of BMI.

**Reply by mdpiper** [reply]
@gregtucker How should we present language specifications?

**Reply by mdpiper** [reply]
@rcabell: NetCDF supports only C, C++, Fortran. But many many unofficial implemetations. Present without support burden.

**Reply by mdpiper** [reply]
@hrajagers Before inviting new languages spec developers to Council, have a waiting period to get interoperability.

**Reply by mdpiper** [reply]
@RolfHut The Octave and R BMI specs were written on the fly, as needed for coupling, so they may need review.

**Comment by mdpiper** [comment]
## Geospatial BMI support

Ask the Council for advice on #99. Should it go through RFC, or possibly be broken into smaller PRs?

**Reply by mdpiper** [reply]
@mcflugen This should be an extension. People won't necessarily need CRS information for a model, so we shouldn't make them include this function.

**Reply by mdpiper** [reply]
This could a part of the grid extension, if we chose to go this way.

**Reply by mdpiper** [reply]
Or, we could add `get_grid_coordinate`, `get_grid_coordinate_names`, and `get_grid_coordinate_units` now (BMI 2.1), then add `get_grid_crs` as an extension.

**Reply by mdpiper** [reply]
@RolfHut Social models have no need for a grid, so it might be good to make grids an extension.

**Reply by mdpiper** [reply]
@peckhams We should be careful with the wording of how the `get_grid_crs` function works. It should return a standard representation of a CRS, not any string.

**Reply by rcabell** [reply]
Models on irregular meshes or polygon elements wouldn't work with this either, +1 for extension

**Reply by mdpiper** [reply]
@peckhams We should make sure the output of this function can effectively be used, e.g., for reprojection.

**Reply by mdpiper** [reply]
@mcflugen How do we word get_var_units? The wording for get_grid_crs could be approached in a similar manner.

**Comment by mdpiper** [comment]
## Post-meeting actions

1. @mcflugen and @mdpiper will write up a plan for extensions and post as an issue. If the Council is happy with the plan, we'll implement it as a PR.
2. @mdpiper will break up #99, making `get_grid_crs` an extension. (Depends on above)
